flipperdevices/flipperzero-firmware
# NFC: fixes & improvements

# What's new

- Fixed reading for Mifare Ultralight tags - checks reading boundary and discards excess.
- Fixed FAST_READ command emulation for Mifare Ultralight - handles requested page range correctly.
- Improved Mifare Ultralight emulation stability - increased log level for some messages, because frequent calls to furi_hal_console_put() somehow directly affects emulation stability (delays due to locks or something, dunno).
- Implemented non-strict PWD_AUTH command emulation for NTAG21x - performs authentication if PACK is not set to default value, otherwise accepts any passwords.
- _Plastic figures and cards with NTAG215 from a well-known Japanese company are now correctly read and emulated_ :smirk:

# Verification 

- Build and download.
- Read some NTAG21n tag:
  - Save, check number of pages in saved file (e.g. NTAG215 should contain pages from 0 to 134 inclusive)
  - Emulate that tag and try FAST_READ command (e.g. 3A 00 00 should return first 4 bytes of UID)
  - Replace PWD and PACK with actual values in saved file, emulate and try PWD_AUTH command:
  ```
  # NTAG215 primer
  Page 133: 00 E6 AC 0C
  Page 134: DE AD 00 00
  
  # Command: 1B 00 E6 AC 0C
  # Response: DE AD (PACK)
  
  # Command: 1B F1 F2 F3 F4
  # Response: 00 (NAK)
  ```

# Checklist (do not modify)
